### PR TITLE
[WIP] Testers refactoring

### PR DESCRIPTION
**Pull request by everzet**
This is a big refactoring of testers towards a singular interface. This should be the last piece of puzzle for 3.1.

This refactoring abstracts all interfaces into two - `Tester` and `ArrangingTester`. In order to support singular interface, `TestContext` and `RunControl` interfaces were introduced to properly encapsulate tester arguments. This refactoring allowed to do a great simplification of hooked and event-dispatching testers.

### Motivation

1. This refactoring should enable very easy fix of #650, #653 
2. It also fixes unnoticed bug in outline examples isolation (they aren't really isolated in 3.0)
3. Much simplified code of all testers, including decorators
4. Preparation for Testwork extraction and release

### BC Breaks

This refactoring effectively removes old tester interfaces, which obviously breaks backwards compatibility. Hence, this PR is in *WIP* state. There will be an additional BC-layer for extensions created prior to 3.1. If all goes well, we will fully support old interfaces for extensions that used them, meaning if you created one for 3.0, in 3.1 it should just work.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,9 +17,9 @@ branches:
 matrix:
   include:
     - php: 5.5
-      env: SYMFONY_VERSION='2.3.*'
+      env: SYMFONY_VERSION='2.5.*' # Remove in July 2015
     - php: 5.5
-      env: SYMFONY_VERSION='2.5.*'
+      env: SYMFONY_VERSION='2.3.*' # LTS - Remove in May 2013
```
Comment (stof) [inline .travis.yml]: the date looks wrong

**Comment by openprivacy** [comment]
This looks good (other than perhaps the date in a comment) and a valuable contribution.

When can I look forward to this being merged and a 3.1 release? (What can I do to help?)

File: src/Behat/Behat/EventDispatcher/Event/GherkinEventFactory.php [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        return $this->exerciseFactory->createBeforeTestedEvent($context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createAfterSetupEvent(TestContext $context, Setup $setup)
+    {
+        switch (true) {
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/EventDispatcher/Event/GherkinEventFactory.php]: Switching on the type of the context suggests generation of the events could be responsibility of the context itself

File: .travis.yml (hunk 1 above)
Comment (fluxsauce) [inline .travis.yml]: As per http://symfony.com/doc/current/contributing/community/releases.html

Version|Feature Freeze|Release|End of Maintenance|End of Life
---|---|---|---|---
2.3|03/2013|05/2013|05/2016 (36 months)|05/2017

**Comment by everzet** [comment]
I still think there might be some merit in this PR, but I think it might have went way overboard with structural changes. I'll keep it open for now, but will not merge it into 3.1.

**Comment by everzet** [comment]
This was too big of a change. When we come closer to 4.0, we'll probably end up doing something similar, but for now - I'm closing this thing. I'll keep branch hanging for future references.

**Pull request closed without merging**